common/lib/dependabot/dependency_updaters/README.md

```diff
@@ -0,0 +1,63 @@
+# Dependency Updater
+
+A *dependency updater* is responsible for updating a dependent package
+according to a set of requirements.
+It provides a minimal interface for languages to add Dependabot support
+by delegating responsibilities to a package manager for that language.
+
+## Public API
+
+Each `Dependabot::DependencyUpdaters` class inherits from `Base` 
+and implements the following methods:
+
+### `#update(dependency:, requirements:)`
+
+Updates a dependency according to the provided requirements.
+This method is called within the root directory of a project.
+
+#### Arguments
+
+| Argument       | Type                     | Description                      |
+| -------------- | ------------------------ | -------------------------------- |
+| `dependency`   | `Dependabot::Dependency` | The package dependency to update |
+| `requirements` | `VersionRange`           | The updated version constraints  |
+
+#### Return
+
+Returns `true` if the dependency was successfully updated 
+to satisfy the new requirements.
+
+#### Raise
+
+This method should raise an exception to communicate underlying errors
+in the package manager.
+
+#### Example
+
+```ruby
+require 'dependabot/dependency_updaters'
+
+# Creating and registering a dependency updater
+
+module Dart
+  class Pub < Dependabot::DependencyUpdaters::Base
+    class Error < StandardError; end
+
+    def update(dependency:, requirements:)      
+      system "dart pub upgrade #{dependency.name} --requirements #{requirements}"
+    rescue => e
+      raise Error.new(e.message)
+    end
+  end
+end
+
+Dependabot::DependencyUpdaters.register("pub", Dart::Pub)
+
+# Updating a dependency
+
+dependency = Dependency.new(package_url: "pkg://pub/http_parser")
+requirements = VersionRange.parse(">=4.0.0")
+
+dependency_updater = Dependabot::DependencyUpdaters.for(dependency.package_manager)
+dependency_updater.update(dependency: dependency, requirements: requirements)
```

Comment:
I'm curious how we'd get the information that we need to feed into a PR, some of could be answered by https://github.com/dependabot/dependabot-core/pull/4542/files#r771474851, but there might be other options

Reply:
Maybe we should include the other (existing or fictional) classes that make up a complete PR in the doc here?

Reply:
Great call. We should definitely think in terms of the full end-to-end functionality.

A few options to consider:

1. Have this method return diff information, as suggested above
2. Add a new required method (say, `#dependencies`); to get a diff, we run once before `#update` and again afterward and compare.
2a. Alternatively, a `#manifest_files` that returns all manifest files in the project, and a `#dependencies(manifest_file:)` that returns dependencies scoped to a particular manifest file.